Wikipedia/Code/WMFAppViewController.m

```diff
@@ -882,7 +887,7 @@ - (void)finishResumingApp {
             if ([error.domain isEqualToString:NSURLErrorDomain]) {
                 return;
             }
-            [self wmf_showReloginFailedPanelIfNecessaryWithTheme:self.theme];
+            [self showLoggedOutPanelIfNeeded];
```

Comment:
`[self showLoggedOutPanelIfNeeded]` should be called regardless of whether or not there's a network error. It also should be called inside of the success block, since there might be a successful non-login while the user is unaware of a previous logout.

Reply:
Also, everything in the above completion block needs to be called as well now that the failure block is working. Everything could be combined into one shared `dispatch_block_t loginAttemptCompletion`

Reply:
Could you explain what you mean by "everything in the above completion block needs to be called as well now that the failure block is working"?

Reply:
Everything in the success completion block - the one declared here `[[WMFAuthenticationManager sharedInstance] attemptLoginWithCompletion:^{`

Reply:
Everything in the success block should be called regardless of whether or not the user is logged in. Since the success block was always being called before, it wasn't an issue. Now that failure is fixed and will be called sometimes, they should share a completion block that does what the success block was doing before, plus `[self showLoggedOutPanelIfNeeded]`

Reply:
But called where? Attempt login takes a completion block and a failure block, then attempt login transforms those blocks into AuthenticationResultHandler. The completion block is called unless it's a failure.

Reply:
Everything in the prior success block needs to be called even if it's a failure, so it should be called in both the success and the failure blocks.
```
dispatch_block_t loginAttemptCompletion = ^{
        [self showLoggedOutPanelIfNeeded];
        [self checkRemoteAppConfigIfNecessary];
        if (!self.reachabilityNotifier) {
...
        }
        [self.periodicWorkerController start];
        [self.savedArticlesFetcher start];
        [self.reachabilityNotifier start];
        self.resumeComplete = YES;
    };

    [[WMFAuthenticationManager sharedInstance] attemptLoginWithCompletion:loginAttemptCompletion failure:^(NSError *error) { loginAttemptCompletion(); }];
```

Alternatively, you could rewrite `attemptLoginWithCompletion` to only have one completion block.

Reply:
Ok, it makes more sense to have one completion then.